trestle/core/rule_parameters_validator.py

```diff
@@ -0,0 +1,125 @@
+
+        args:
+            by_components: A list of by components.
+            components: Components defined at System implementation level.
+            cat_int: Instance of catalog interface with controls catalog loaded.
+            imp_requirement: Current implemented requirement.
+            statement: Statement if iterating by statements.
+        """
+        for by_component in as_list(by_components):
+            component = get_item_from_list(components, by_component.component_uuid, lambda x: x.uuid)
+            # first adds implmented requirement set params
+            if not statement:
+                self._add_shared_param(imp_requirement, imp_requirement.control_id, cat_int, component.title)
+            # then  adds by components set params per requirement
+            self._add_shared_param(by_component, imp_requirement.control_id, cat_int, component.title)
+
+    def _add_shared_param(
+        self, item: TypeWithSetParams, control_id: str, cat_int: CatalogInterface, comp_name: str = ''
+    ) -> None:
+        """
+        Add a rule shared parameter to the rule shared parameters list.
+
+        args:
+            item: Generic item to iterate over parameters.
+            control_id: Current control id.
+            cat_int: Instance of catalog interface with controls catalog loaded.
+            comp_name: Component name to save.
+        """
+        for set_param in as_list(item.set_parameters):
+            control = cat_int.get_control_by_param_id(set_param.param_id)
+            if not control:
+                deep_set(self._rule_param_values_dict, [set_param.param_id, comp_name, control_id], set_param.values)
+
+    def model_is_valid(
+        self, model: TopLevelOscalModel, quiet: bool, trestle_root: Optional[pathlib.Path] = None
+    ) -> bool:
+        """
+        Test if the model is valid.
+
+        args:
+            model: A top level OSCAL model.
+            quiet: Don't report msgs unless invalid.
+            trestle_root: Trestle root path.
+
+        returns:
+            True (valid) if the model's rule parameter values are the same across controls.
+        """
+        # verify if model type is either an SSP of a Component Definition
+        if not isinstance(model, SystemSecurityPlan):
+            return True
+
+        components = as_list(model.system_implementation.components)
+        if not model.import_profile.href:
+            logger.info(f'INVALID: Model {model.metadata.title} has no referenced profile')
+            return False
+        profile_catalog = ProfileResolver().get_resolved_profile_catalog(trestle_root, model.import_profile.href)
+        catalog_interface = CatalogInterface(profile_catalog)
+        # iterate by each implemented requirement defined
+        for imp_req in as_list(model.control_implementation.implemented_requirements):
+            # iterate by each by_component in each implemented requirement
```

Comment:
you don't need as_list here because imp_reqs are required

Reply:
done